oas_docs/bundle.json

```diff
@@ -1515,6 +1534,27 @@
                                       "description": "Total time spent performing Elasticsearch searches as measured by Kibana; includes network latency and time spent serializing or deserializing the request and response.",
                                       "nullable": true,
                                       "type": "number"
+                                    },
+                                    "unfilled_gaps_ms": {
+                                      "additionalProperties": false,
+                                      "nullable": true,
+                                      "properties": {
+                                        "last_1d": {
+                                          "type": "number"
+                                        },
+                                        "last_3d": {
+                                          "type": "number"
+                                        },
+                                        "last_7d": {
+                                          "type": "number"
+                                        }
+                                      },
+                                      "required": [
+                                        "last_1d",
+                                        "last_3d",
+                                        "last_7d"
+                                      ],
+                                      "type": "object"
```

Comment:
I'm concerned that storing these metrics as part of the last run metrics will force us to recompute them on every rule execution by querying the event log - even on rule executions that are not filling or detecting gaps. After a gap, we'll have to update these metrics on every rule execution for the next 7 days, which means a lot of extra event log queries to ES. @mikecote found in the 10x alerting scale effort that reducing the number of requests made by each rule execution is very important for minimizing load on ES and running the total number of rules we want. We will get better performance with an approach that only queries the event log for gaps on rule executions that are filling or detecting a gap.

What's the reasoning for not storing an array of date ranges that represent the list of unfilled gaps? The list will be small and runtime fields could compute the last_1d, last_3d, last_7d values at query time.

We need to either show how this data structure could be updated at rule execution time without making event log queries covering a large time range (last 7 days) or pick a different data structure.

Reply:
I go with use this approach, after [that discussion](https://elastic.slack.com/archives/C06D75UHFD3/p1730383760121909) in the chat. And idea was not to store potentially thousands  of intervals in SO. 

Current idea, that after each rule execution we will do 1 extra ES request, to calculate information about gaps for last 1/3/7 days.

For rule which has no gaps, those request should be pretty fast, as query return 0 documents.
For rule with gaps, and potentially, 1000 of documents I believe aggregations should be still pretty fast.

Considering how many es request we already do on rule execution, I don't think it will dramatically decrease performance.

We also can't only calculate those data only manual rule run, because there can be no manual rule run, and after 1 day, information about gaps can be old/ not actual.

@ymao1 @mikecote Can you give your feedback on those 2 approaches?

As alternative approach to both runtime field and storing last 1/3/7 days, we can consider to not put this data into rule monitoring table, and create another UI, for rule with gaps.

Reply:
Do we know any metrics on how often we currently see unfilled gaps? We were thinking that if the current rule execution returns no gaps and `unfilled_gaps.*` is currently 0, we could skip any queries against the event log and we were hoping this would be the majority of cases during rule execution.

Reply:
[This link](https://overview.elastic-cloud.com/s/security-solution/app/r/s/gLggF) shows how often we're seeing gaps with our current method of gap detection, although that includes times when a rule was disabled and re-enabled later. Long term, I expect that we'll exclude disable -> re-enable from counting as a gap but **include** other types of rule execution failures as gaps, so another good metric to look at is search failures: https://overview.elastic-cloud.com/s/security-solution/app/r/s/nBfmU.

> Considering how many es request we already do on rule execution, I don't think it will dramatically decrease performance.

One request alone may not dramatically decrease performance, but we need to be careful about introducing more requests. This is one reason we haven't introduced other features that could be useful like count of documents excluded by exceptions, total count of documents in the time range covered by a rule, etc. We're going to be pushing to reduce the number of queries made by each rule execution as much as possible. Some rule types like IM are naturally expensive and make a lot of queries - we may not be able to reduce those much. But EQL and KQL are by far the most common and often make a single core search request to ES that **must** be made on each rule execution, and all the other work to set up for that query (field caps, exception fetching, privilege checks) are candidates for caching. Ideally we want to get the average number of ES requests per rule execution as close to 1 as possible.

> For rule with gaps, and potentially, 1000 of documents I believe aggregations should be still pretty fast.

I don't think we can assume that. Gaps are not independent across rules, when one rule has a gap it's likely that other rules will also have gaps. In a system running 3200+ rules per minute, we will see 3200+ extra event log queries per minute after a service interruption and the extra query load will persist for 7 days or until the gaps are filled. In this case it's particularly risky because all those queries from all rules are hitting the same index, which then means we need to ensure the event log has enough shards to handle the sustained extra load.